enhancements/api-review/gcp_user_defined_labels.md

```diff
@@ -0,0 +1,566 @@
+  in the install config. The tags are added by installer and the in-cluster operators while creating GCP resources.
+
+### Non-Goals
+
+- Support for update or delete operation on labels or tags after creation of GCP resource.
+
+## Proposal
+
+***GCP Labels*** 
+
+A label of the form `kubernetes-io-cluster-<cluster_id>:owned` will be added to every
+resource created by OpenShift to enable administrator to differentiate the resources
+created for OpenShift cluster. An administrator is not allowed to add or modify the label
+having the prefix `kubernetes-io` or `openshift-io` in the key. The same can be found
+applied to other cloud platform resources which supports labels/tags for ex: AWS, Azure.
+
+New `userLabels` field will be added to `platform.gcp` of install-config for the user
+to define the labels to be added to the resources created by installer and in-cluster operators.
+
+If `platform.gcp.userLabels` in the install-config has any labels defined, then these labels
+will be added to all the GCP resources created by OpenShift, provided all the defined
+labels meet all the below conditions or else the cluster creation will fail.
+1. A label key and value must have minimum of 1 character and can have maximum of 63 characters.
+2. A label key and value must contain only lowercase letters, numeric characters,
+   underscores, and dashes.
+3. A label key must start with a lowercase letter.
+4. Each resource can have multiple labels, up to a maximum of 32.
+   - GCP supports a maximum of 64 labels per resource. Restricting the number of
+    user defined labels to 32 and reserving 32 for Openshift's internal use.
+
+All in-cluster operators that create GCP resources (Cluster Infrastructure ,Storage) will
+apply these labels during resource creation.
+
+The userLabels field is intended to be set at install time and is considered immutable.
+Components that respect this field must only ever add labels that they retrieve from this
+field to cloud resources, they must never remove labels from the existing underlying cloud
+resource even if the labels are removed from this field(despite it being immutable).
+
+If the userLabels field is changed post-install, there is no guarantee about how an
+in-cluster operator will respond to the change. Some operators may reconcile the
+change and update the labels on the GCP resource. Some operators may ignore the change.
+However, if labels are removed from userLabels, the label will not be removed from the
+GCP resource.
+
+***GCP Tags*** 
+
+Unlike labels, Tags are not metadata of a resource but a resource itself. Tags are key-value
+pairs. Tag keys, values, and bindings are all discrete resources. A tag key resource must be
+created under an organization resource. A tag value created for each key is a list of values
+that can be assigned to the key. A maximum of 1000 keys can be created under a given
+organization and there can be a total of 1000 values created for each key.
```

Comment:
Are these absolute limits or can they be increased by a request to GCP? How will we ensure customers understand this since it will put a limit on the number of clusters that can be created? 
Do we know how many keys an openshift cluster will consume by default?

Reply:
Yeah, those are current limits for tags. I will check if those can be extended, but I didn't come across about the same in their docs.

Since these tags must already exist before cluster creation, openshift won't be adding any tags.

Reply:
Can the tags be shared among clusters, or are they likely to be unique per cluster? If they are shared then this probably isn't an issue

Reply:
Yeah, it can be shared among clusters. Any tag added to a parent resource will be inherited by all the child resources. If a tag is added to a project, then all resources created for all the clusters under this project will inherit the tags. And each resource can be added with new tags and inherited tags can be modified.

Reply:
Ok, in that case I'm not too worried about the limits, but we should make sure they are documented when the feature is documented